File: App/app/src/main/java/com/example/runpal/models/Event.kt
package com.example.runpal.models

import android.net.Uri
import com.google.android.gms.maps.model.LatLng

/**
 * Represents a running event.
 *
 * @param _id The event identifier.
 * @param time The UNIX timestamp of event start.
 * @param followers Number of followers.
 * @param following True if the logged in user is following the event.
 */
data class Event(
    val _id: String = "",
    val name: String = "",
    val description: String = "",
    var image: String = "",
    val time: Long = 0L,
    val distance: Double = 0.0,
    val path: List<LatLng>? = null,
    val tolerance: Double? = null,
    val followers: Long = 0L,
    val following: Boolean = false
) {

    enum class Status {
        UPCOMING, CURRENT, PAST
    }

    var imageUri: Uri
        get() = Uri.parse(image)
        set(value) {image = value.toString()}

    val status: Status
        get() {
            val now = System.currentTimeMillis()
            if (now < time - 60*60*1000) return Status.UPCOMING
            else if (now < time + 60*60*1000) return Status.CURRENT
            else return Status.PAST
        }
}

File: Server/src/models/Event.ts
import { ObjectId } from 'mongodb'
import mongoose, {InferSchemaType} from 'mongoose'

const latLngSchema = new mongoose.Schema({
    latitude: {
        type: Number,
        required: true
    },
    longitude: {
        type: Number,
        required: true
    }
})
const eventSchema = new mongoose.Schema({
    name: {
        type: String,
        required: true
    },
    description: {
        type: String,
        required: true,
        default: ""
    },
    image: {
        type: String,
        required: true
    },
    time: {
        type: Number,
        required: true
    },
    distance: {
        type: Number,
        required: true
    },
    path: {
        type: [latLngSchema],
        required: false,
        default: null
    },
    tolerance: {
        type: Number,
        required: false,
        default: null
    },
    followers: {
        type: [mongoose.Schema.Types.ObjectId],
        required: true,
        default: []
    }
})

export type Event = InferSchemaType<typeof eventSchema> & {_id?: ObjectId}
export type EventRanking = {
    user: mongoose.Schema.Types.ObjectId,
    name: string,
    last: string,
    time: number,
    disqualified: boolean
}
export default mongoose.model("eventModel", eventSchema, "events")